### DISCUSSION TITLE
NotImplementedError: None 问题

### DISCUSSION THREAD

**Discussion by Redamancy-520** [fangwei123456/SpikingFlow#202] [Q&A] [answered]
在运行 conv_fashion_mnist.main() 代码过程中报错 NotImplementedError: None，想问一下如何更正谢谢！

**Comment by fangwei123456** [comment, marked as answer]
提供一下可以复现的代码和运行参数

**Reply by Redamancy-520** [reply]
就是您写的教程中的1.6节 时间驱动：使用卷积snn识别fashion-mnist 1.6.3部分

from spikingjelly.clock_driven.examples import conv_fashion_mnist

conv_fashion_mnist.main()

**Reply by fangwei123456** [reply]
不能直接这样运行，要用命令行传入参数。`python -m spikingjelly.clock_driven.examples.conv_fashion_mnist -h`

**Reply by fangwei123456** [reply]
![image](https://user-images.githubusercontent.com/38352109/162107934-70309789-d643-478d-8e5a-8b9ae6ef9de4.png)

**Reply by Redamancy-520** [reply]
您好，在我这版本的指导教程里好像并没有提及，而且直接这样运行1.1，1.2，1.5节的代码也是可行的。我看了一下，系统会输入默认的参数值。
![屏幕截图 2022-04-07 104403](https://user-images.githubusercontent.com/65457793/162109923-a7573d14-ce8e-495d-a01e-13e02469e0ec.png)

**Reply by fangwei123456** [reply]
上古版本是这样，实时输入参数，后来改成了从命令行传入。你报的这个错误应该是老版教程却使用了新版代码。

**Reply by Redamancy-520** [reply]
谢谢您，我明白了。另外，在尝试您的提供的方法之后，我看到系统提示Ninja is required to load C++ extensions，但是我还没有安装ninja。在安装ninja之后系统又会报错 Error checking compiler version for cl: [WinError 2]  系统找不到指定的文件。这个是系统问题吗？

**Reply by fangwei123456** [reply]
升级SJ到最新版本试试，从GitHub源码安装

**Reply by Redamancy-520** [reply]
好的，我试试，再次感谢

**Comment by Redamancy-520** [comment]
您好，使用git命令尝试从GitHub源码安装SJ系统会一直报错 [Errno 13] Permission denied，发现尝试写入系统的只读文件夹且该文件夹无法更改只读模式，想问一下解决方法或者别的安装方式，谢谢您

**Reply by fangwei123456** [reply]
在conda创建的虚拟环境里面装

**Reply by Redamancy-520** [reply]
您好，在anaconda虚拟环境中按照您的方式运行之后依然会报错。

D:\Anaconda3\envs\snn\lib\site-packages\torch\utils\cpp_extension.py:322: UserWarning: Error checking compiler version for cl: [WinError 2] 系统找不到指定的文件。
warnings.warn(f'Error checking compiler version for {compiler}: {error}')
信息: 用提供的模式无法找到文件。
spikingjelly.clock_driven.spike_op: Command '['where', 'cl']' returned non-zero exit status 1.
FileNotFoundError: [WinError 3] 系统找不到指定的路径。: './logs\\T_4_b_128_SGD_lr_0.1_CosALR_64_amp'

我看到您之前解答过类似问题，因此在系统的环境变量中的Path已经加入了cl.exe，但是还是报一样的错。

**Reply by fangwei123456** [reply]
在命令提示符输入cl.exe，看是否能正常输出
![image](https://user-images.githubusercontent.com/38352109/162193933-d88acc84-9d05-4000-81ff-702e4423c43d.png)
在新版中没有cl只会警告，不会报错。
`FileNotFoundError: [WinError 3] 系统找不到指定的路径。: './logs\T_4_b_128_SGD_lr_0.1_CosALR_64_amp'`这个是logs文件夹不存在，需要新建logs文件夹，这样日志才能保存在里面。

**Reply by Redamancy-520** [reply]
输入cl.exe文件后可以正常输出

**Reply by fangwei123456** [reply]
那就先忽略那些吧，windows平台上一般也用不到`spikingjelly.clock_driven.spike_op`

**Reply by Redamancy-520** [reply]
好，谢谢您，我去看看

**Reply by Redamancy-520** [reply]
您好，logs文件夹新建在anaconda环境里吗，我在虚拟环境中的文件夹新建logs依然显示FIleNotFound

**Reply by fangwei123456** [reply]
在你运行这个代码的根目录

**Reply by Redamancy-520** [reply]
您好，修改之后又出现了新的问题

C:\Users\WuRap\AppData\Local\torch_extensions\torch_extensions\Cache\py39_cu113\cpp_wrapper\main.cpp(4): error C2065: “cudnn_convolution_backward”: 未声明的标识符
C:\Users\WuRap\AppData\Local\torch_extensions\torch_extensions\Cache\py39_cu113\cpp_wrapper\main.cpp(5): error C2065: “cudnn_convolution_backward_input”: 未声明的标识符
C:\Users\WuRap\AppData\Local\torch_extensions\torch_extensions\Cache\py39_cu113\cpp_wrapper\main.cpp(6): error C2065: “cudnn_convolution_backward_weight”: 未声明的标识符
ninja: build stopped: subcommand failed.

是否还是与cl.exe的问题有关？之前也是ninja包的报错

**Reply by fangwei123456** [reply]
看报错信息是没有安装cudnn

**Reply by Redamancy-520** [reply]
您好，我看了一下，我这边是已经安装了cudnn的，为了防止版本不一致重装看一次之后依然报错。
![image](https://user-images.githubusercontent.com/65457793/162395825-8f044180-48c0-46ce-bd66-aef77bb52ac7.png)

**Reply by fangwei123456** [reply]
看一下torch
![image](https://user-images.githubusercontent.com/38352109/162396100-cec8c83e-5083-4e56-a981-4a9ad4892004.png)

**Reply by Redamancy-520** [reply]
![image](https://user-images.githubusercontent.com/65457793/162396303-3da1d11f-f4af-449c-b320-5e5f3559e04e.png)

**Reply by fangwei123456** [reply]
![image](https://user-images.githubusercontent.com/38352109/162396631-f52b0317-60c0-48fc-bf77-7c3c89420179.png)

可能还是cl.exe的问题，删除`C:\Users\WuRap\AppData\Local\torch_extensions\torch_extensions\Cache\py39_cu113\cpp_wrapper`文件夹试试

**Reply by Redamancy-520** [reply]
好的，我看了一下ninja版本也是对的，我去试试，谢谢您！

**Reply by Redamancy-520** [reply]
跑了一次后直接报错
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling `cublasCreate(handle)`

之前是一直卡在ninja: build stopped: subcommand failed，会一直循环。我不确定是不是电脑的问题？因为我这台电脑运行内存只有8G，之前因为内存报错过，我手动修改了c盘的虚拟内存

**Reply by fangwei123456** [reply]
猜测是windows下的依赖问题，换linux跑吧

**Reply by Redamancy-520** [reply]
是的，我也觉得，我去linux下试试，辛苦您了

**Comment by Redamancy-520** [comment]
另外，我在运行教程中别的部分的代码时也发现了几个问题。
在ann2snn部分，运行代码“cnn_mnist.main()”时 会报错：ImportError: cannot import name 'parser'，我看到教程中写道parser类定义在__init()__中，但是打开__init()__文件中也没有看到。

在强化学习DQN部分，运行 “Spiking_DQN_state.train(use_cuda=False, model_dir='./model/CartPole-v0', log_dir='./log', env_name='CartPole-v0', hidden_size=256, num_episodes=500, seed=1)” 代码启动训练时会报错 TypeError: SummaryWriter.__init__() got an unexpected keyword argument 'logdir'。我查询了一下，应该是与tensorboard的接口参数改变导致，log_dir应该改为logdir。我尝试自己修改，但是涉及参数有点多，还是无法成功运行代码。

十分抱歉麻烦您这么久，我刚刚进入赵老师组里，正在熟悉SJ平台，谢谢您！

**Reply by fangwei123456** [reply]
![image](https://user-images.githubusercontent.com/38352109/162195151-7ad206a1-f6ee-4d70-8c9c-e66b7464e2cb.png)
![image](https://user-images.githubusercontent.com/38352109/162194985-6298c68c-4558-45a2-b5a7-0fcc99b10119.png)
@DingJianhao ann2snn的教程需要修复一下了

**Reply by fangwei123456** [reply]
DQN那个bug修复了，需要用git pull更新代码并重新安装

**Reply by Redamancy-520** [reply]
好的，谢谢您

**Reply by fangwei123456** [reply]
赵老师是哪位老师？

**Reply by Redamancy-520** [reply]
赵地老师

**Reply by fangwei123456** [reply]
好的，听说你们准备写STDP之类的学习规则？如果写的话可以参考https://github.com/fangwei123456/spikingjelly/issues/76

**Reply by Redamancy-520** [reply]
好的，辛苦您啦

**Reply by DingJianhao** [reply]
我下个季度来修改，最近要综测😭

**Reply by fangwei123456** [reply]
好，忙完再改吧

**Reply by Redamancy-520** [reply]
您好，DQN的bug更改之后还有新的问题

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "c:\users\wurap\spikingjelly\spikingjelly\clock_driven\examples\Spiking_DQN_state.py", line 101, in train
    policy_net = DQSN(n_states, hidden_size, n_actions, T).to(device)
  File "c:\users\wurap\spikingjelly\spikingjelly\clock_driven\examples\Spiking_DQN_state.py", line 60, in __init__
    NonSpikingLIFNode(tau=2.0)
NameError: name 'NonSpikingLIFNode' is not defined

**Reply by fangwei123456** [reply]
这个错误修复了，更新代码试试

**Reply by Redamancy-520** [reply]
您好，我更新代码之后报错

Traceback (most recent call last):
  File "C:\Users\WuRap\PycharmProjects\pythonProject\RL_DQN\main.py", line 3, in <module>
    Spiking_DQN_state.train(use_cuda=False, model_dir='./model/CartPole-v0', log_dir='./log', env_name='CartPole-v0', hidden_size=256, num_episodes=500, seed=1)
  File "C:\Users\WuRap\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\spikingjelly\clock_driven\examples\Spiking_DQN_state.py", line 192, in train
    torch.save(policy_net.state_dict(), max_pt_path)
  File "C:\Users\WuRap\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\torch\serialization.py", line 377, in save
    with _open_file_like(f, 'wb') as opened_file:
  File "C:\Users\WuRap\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\torch\serialization.py", line 231, in _open_file_like
    return _open_file(name_or_buffer, mode)
  File "C:\Users\WuRap\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\torch\serialization.py", line 212, in __init__
    super(_open_file, self).__init__(open(name, mode))
FileNotFoundError: [Errno 2] No such file or directory: './model/CartPole-v0\\policy_net_256_max.pt'

我是需要下载一些数据还是因为数据路径的问题？

**Reply by fangwei123456** [reply]
路径问题，应该是文件夹`./model/CartPole-v0`不存在导致保存模型失败

**Reply by Redamancy-520** [reply]
已经解决了，辛苦您了！
